### PR TITLE
2794 retention people

### LINKED ISSUE #2794: Bug with CH persons double click on retention table
## Bug description

We have an issue in which the table of users for each cohort does not contain the same number of users as the table reports.

### Example
([Source](https://app.posthog.com/insights?startEntity=%7B%22actions%22%3A%5B%5D%2C%22events%22%3A%5B%7B%22id%22%3A%22user%20signed%20up%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%2C%22name%22%3A%22user%20signed%20up%22%7D%5D%2C%22new_entity%22%3A%5B%5D%7D&returningEntity=%7B%22events%22%3A%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%7D%5D%7D&retentionType=retention_first_time&date_to=2020-04-05T12%3A59%3A36.001Z&period=w&display=ActionsTable&properties=%5B%7B%22key%22%3A%22id%22%2C%22value%22%3A71%2C%22type%22%3A%22cohort%22%7D%2C%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22www.posthog.com%22%2C%22operator%22%3A%22not_icontains%22%2C%22type%22%3A%22event%22%7D%2C%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22https%3A%2F%2Fposthog.com%22%2C%22operator%22%3A%22not_icontains%22%2C%22type%22%3A%22event%22%7D%2C%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22localhost%22%2C%22operator%22%3A%22not_icontains%22%2C%22type%22%3A%22event%22%7D%2C%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%22127.0.0.1%22%2C%22operator%22%3A%22not_icontains%22%2C%22type%22%3A%22event%22%7D%2C%7B%22key%22%3A%22%24current_url%22%2C%22value%22%3A%220.0.0.1%22%2C%22operator%22%3A%22not_icontains%22%2C%22type%22%3A%22event%22%7D%5D&insight=RETENTION))

**Table shows 6 users in cohort:**

**Persons list shows a lot more than 6 users in cohort:**

## Expected behavior
The persons list and the retention table should match.

## How to reproduce

See description above.

## Environment

Cloud, Clickhouse.

## Additional context

Currently blocking internal personas analysis.

### LINKED ISSUE #3081: Retention trends graph people
## Bug description

*Please describe.*  
- some datapoints show 0 but people will appear in modal

*If this affects the front-end, screenshots would be of great help.*  

## Expected behavior
- amount of people and datapoint should be the same

## How to reproduce

1. demo data clickhouse instance
2. go to retention trends for a day with some people

## Environment

- [ ] PostHog Cloud
- [ ] self-hosted PostHog, version/commit: _please provide_

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- fixes #2794 
- fixes #3081 
*If this affects the frontend, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

**Comment by EDsCODE** [comment]
waiting on #3091 for types

**Comment by EDsCODE** [comment]
@timgl bump review. few bugs are waiting on this

**Comment by macobo** [comment]
Q: Will this fix https://sentry.io/organizations/posthog/issues/2095798964/?project=1899813&query=is%3Aunassigned+is%3Aunresolved?

File: frontend/src/scenes/retention/RetentionTable.tsx [hunk 1]
```diff
@@ -133,25 +129,24 @@ export function RetentionTable({
...
                                                                     )
                                                                 </span>
                                                             )}
                                                         </td>
                                                     ))}
                                             </tr>
-                                            {people.result &&
-                                                people.result.map((personAppearances: RetentionTableAppearanceType) => (
+                                            {people.result?.detail &&
+                                                (people.result.detail as any[]).map((personAppearances) => (
```
Comment (macobo) [inline frontend/src/scenes/retention/RetentionTable.tsx:149]: Is the `as` cast here needed? I think it isn't. At least it's removing type safety here.

File: posthog/queries/retention.py [hunk 2]
```diff
@@ -280,14 +280,16 @@ def _retrieve_people_in_period(self, filter: RetentionFilter, team: Team):
 
     def process_people_in_period(
         self, filter: RetentionFilter, vals, people_dict: Dict[str, ReturnDict]
-    ) -> List[Dict[str, Any]]:
+    ) -> Dict[str, Any]:
         marker_length = filter.total_intervals
         result = []
+        appearance_totals = [0 for _ in range(marker_length)]
         for val in vals:
```
Comment (macobo) [inline posthog/queries/retention.py:287]: I was having a lot of difficulty unpacking this code.  Nice way to make this more readable:

```
for person_id, _, appearances_array in vals:
```

**Review by macobo — APPROVED** [review]
Not 100% sure I understand this code fully - the query side of things is :magic:, but looks to be correct-ish.

**Review by macobo — APPROVED** [review]
Oops I caused a merge conflict

File: posthog/queries/retention.py (hunk 2 above)
Comment (EDsCODE) [reply posthog/queries/retention.py]: yup def better

File: frontend/src/scenes/retention/RetentionTable.tsx (hunk 1 above)
Comment (EDsCODE) [reply frontend/src/scenes/retention/RetentionTable.tsx]: not sure what that was about. I think i was trying something and it got left in the commit

**Pull request merged**